File: ViewProfit/table_portfolio.py
# -*- coding: utf-8 -*-


import os

import numpy as np
from PySide2.QtCharts import QtCharts
from PySide2.QtCore import QDateTime, QLocale, QObject, Qt, Signal
from PySide2.QtGui import QColor
from PySide2.QtSql import QSqlQuery, QSqlTableModel
from PySide2.QtUiTools import QUiLoader
from PySide2.QtWidgets import (QFrame, QGraphicsDropShadowEffect, QPushButton,
                               QTableView)

from ViewProfit.model_portfolio import ModelPortfolio


class TablePortfolio(QObject):
    new_mouse_coords = Signal(object,)

    def __init__(self, app):
        QObject.__init__(self)

        self.app = app
        self.db = app.db
        self.chart1 = app.chart1
        self.chart2 = app.chart2
        self.name = "portfolio"

        self.module_path = os.path.dirname(__file__)

        self.model = ModelPortfolio(self, app.db)

        self.model.setTable(self.name)
        self.model.setEditStrategy(QSqlTableModel.OnManualSubmit)
        self.model.setSort(1, Qt.SortOrder.DescendingOrder)

        currency = QLocale().currencySymbol()

        self.model.setHeaderData(1, Qt.Horizontal, "Date")
        self.model.setHeaderData(2, Qt.Horizontal, "Total Contribution " + currency)
        self.model.setHeaderData(3, Qt.Horizontal, "Real Bank Balance " + currency)
        self.model.setHeaderData(4, Qt.Horizontal, "Real Return " + currency)
        self.model.setHeaderData(5, Qt.Horizontal, "Real Return %")

        self.model.select()

        loader = QUiLoader()

        self.main_widget = loader.load(self.module_path + "/ui/table_portfolio.ui")

        table_cfg_frame = self.main_widget.findChild(QFrame, "table_cfg_frame")
        self.table_view = self.main_widget.findChild(QTableView, "table_view")
        button_calculate = self.main_widget.findChild(QPushButton, "button_calculate")

        self.table_view.setModel(self.model)

        self.table_view.setColumnHidden(0, True)  # do no show the id column

        # effects

        table_cfg_frame.setGraphicsEffect(self.card_shadow())
        button_calculate.setGraphicsEffect(self.button_shadow())

        # signals

        button_calculate.clicked.connect(self.calculate)

    def button_shadow(self):
        effect = QGraphicsDropShadowEffect(self.main_widget)

        effect.setColor(QColor(0, 0, 0, 100))
        effect.setXOffset(1)
        effect.setYOffset(1)
        effect.setBlurRadius(5)

        return effect

    def card_shadow(self):
        effect = QGraphicsDropShadowEffect(self.main_widget)

        effect.setColor(QColor(0, 0, 0, 100))
        effect.setXOffset(2)
        effect.setYOffset(2)
        effect.setBlurRadius(5)

        return effect

    def calculate(self):
        list_dates = set()

        for table_dict in self.app.tables:
            t = table_dict['object']
            table_type = table_dict['type']

            # making sure all the latest data was saved to the database

            t.model.submitAll()

            if table_type == "investment":
                query = QSqlQuery(self.db)

                query.prepare("select distinct date from " + t.name + " order by date")

                if query.exec_():
                    while query.next():
                        list_dates.add(query.value(0))
                else:
                    print(t.model.lastError().text())

        list_dates = list(list_dates)

        list_dates.sort()

        for date in list_dates:
            total_contribution = 0.0
            real_bank_balance = 0.0
            real_return = 0.0
            real_return_perc = 0.0

            for table_dict in self.app.tables:
                t = table_dict['object']
                table_type = table_dict['type']

                if table_type == "investment":
                    query = QSqlQuery(self.db)

                    """
                        432000 seconds = 5 days. If the time difference is smaller than that we consider that the dates
                        refer to the same month
                    """

                    query.prepare("select distinct date,total_contribution,real_bank_balance,real_return," +
                                  "real_return_perc from " + t.name + " where abs(date - ?) < 432000")

                    query.addBindValue(date)

                    if query.exec_():
                        while query.next():
                            total_contribution += query.value(1)
                            real_bank_balance += query.value(2)
                            real_return += query.value(3)
                            real_return_perc += query.value(4)
                    else:
                        print(t.model.lastError().text())

        # self.recalculate_columns()

        # self.show_chart()

    def recalculate_columns(self):
        self.calculate_total_contribution()

        self.qsettings.beginGroup(self.name)

        income_tax = float(self.qsettings.value("income_tax", 0.0))

        self.qsettings.endGroup()

        for n in range(self.model.rowCount()):
            total_contribution = self.model.record(n).value("total_contribution")

            gross_return = self.model.record(n).value("bank_balance") - total_contribution

            real_return = gross_return * (1.0 - 0.01 * income_tax)

            rec = self.model.record(n)

            rec.setGenerated("gross_return", True)
            rec.setGenerated("gross_return_perc", True)
            rec.setGenerated("real_return", True)
            rec.setGenerated("real_return_perc", True)
            rec.setGenerated("real_bank_balance", True)

            rec.setValue("gross_return", float(gross_return))
            rec.setValue("real_return", float(real_return))
            rec.setValue("real_bank_balance", float(total_contribution + real_return))

            if total_contribution > 0:
                rec.setValue("gross_return_perc", float(100 * gross_return / total_contribution))
                rec.setValue("real_return_perc", float(100 * real_return / total_contribution))

            self.model.setRecord(n, rec)

    def calculate_total_contribution(self):
        list_v = []

        for n in range(self.model.rowCount()):
            list_v.append(self.model.record(n).value("contribution"))

        if len(list_v) > 0:
            list_v.reverse()

            cum_v = np.cumsum(np.array([list_v]))

            cum_v = cum_v[::-1]

            for n in range(self.model.rowCount()):
                rec = self.model.record(n)

                rec.setGenerated("total_contribution", True)

                rec.setValue("total_contribution", float(cum_v[n]))

                self.model.setRecord(n, rec)

    def make_chart1(self):
        self.chart1.setTitle(self.name)

        series0 = QtCharts.QLineSeries()
        series1 = QtCharts.QLineSeries()

        series0.setName("Total Contribution")
        series1.setName("Real Bank Balance")

        series0.hovered.connect(self.on_hover)
        series1.hovered.connect(self.on_hover)

        vmin, vmax = None, None

        for n in range(self.model.rowCount()):
            qdt = QDateTime.fromString(self.model.record(n).value("date"), "dd/MM/yyyy")

            epoch_in_ms = qdt.toMSecsSinceEpoch()

            total_contribution = self.model.record(n).value("total_contribution")
            real_bank_balance = self.model.record(n).value("real_bank_balance")

            if vmin:
                vmin = min(vmin, total_contribution)
                vmin = min(vmin, real_bank_balance)
            else:
                vmin = total_contribution

            if vmax:
                vmax = max(vmax, total_contribution)
                vmax = max(vmax, real_bank_balance)
            else:
                vmax = total_contribution

            series0.append(epoch_in_ms, total_contribution)
            series1.append(epoch_in_ms, real_bank_balance)

        self.chart1.addSeries(series0)
        self.chart1.addSeries(series1)

        axis_x = QtCharts.QDateTimeAxis()
        axis_x.setTitleText("Date")
        axis_x.setFormat("dd/MM/yyyy")
        axis_x.setLabelsAngle(-10)

        axis_y = QtCharts.QValueAxis()
        axis_y.setTitleText(QLocale().currencySymbol())
        axis_y.setLabelFormat("%.2f")
        axis_y.setRange(0.9 * vmin, 1.2 * vmax)

        self.chart1.addAxis(axis_x, Qt.AlignBottom)
        self.chart1.addAxis(axis_y, Qt.AlignLeft)

        series0.attachAxis(axis_x)
        series0.attachAxis(axis_y)

        series1.attachAxis(axis_x)
        series1.attachAxis(axis_y)

    def make_chart2(self):
        self.chart2.setTitle(self.name)

        series0 = QtCharts.QLineSeries()

        series0.setName("Real Return")

        series0.hovered.connect(self.on_hover)

        vmin, vmax = None, None

        for n in range(self.model.rowCount()):
            qdt = QDateTime.fromString(self.model.record(n).value("date"), "dd/MM/yyyy")

            epoch_in_ms = qdt.toMSecsSinceEpoch()

            real_return_perc = self.model.record(n).value("real_return_perc")

            if vmin:
                vmin = min(vmin, real_return_perc)
            else:
                vmin = real_return_perc

            if vmax:
                vmax = max(vmax, real_return_perc)
            else:
                vmax = real_return_perc

            series0.append(epoch_in_ms, real_return_perc)

        self.chart2.addSeries(series0)

        axis_x = QtCharts.QDateTimeAxis()
        axis_x.setTitleText("Date")
        axis_x.setFormat("dd/MM/yyyy")
        axis_x.setLabelsAngle(-10)

        axis_y = QtCharts.QValueAxis()
        axis_y.setTitleText("%")
        axis_y.setLabelFormat("%.2f")

        self.chart2.addAxis(axis_x, Qt.AlignBottom)
        self.chart2.addAxis(axis_y, Qt.AlignLeft)

        series0.attachAxis(axis_x)
        series0.attachAxis(axis_y)

        # add benchmarks

        for table_dict in self.app.tables:
            if table_dict['type'] == "benchmark":
                t = table_dict['object']

                series = QtCharts.QLineSeries()

                series.setName(table_dict['name'])

                series.hovered.connect(self.on_hover)

                for n in range(t.model.rowCount()):
                    qdt = QDateTime.fromString(t.model.record(n).value("date"), "dd/MM/yyyy")

                    epoch_in_ms = qdt.toMSecsSinceEpoch()

                    accu = t.model.record(n).value("accumulated")

                    vmin = min(vmin, accu)
                    vmax = max(vmax, accu)

                    series.append(epoch_in_ms, accu)

                self.chart2.addSeries(series)

                series.attachAxis(axis_x)
                series.attachAxis(axis_y)

        axis_y.setRange(0.9 * vmin, 1.2 * vmax)

    def show_chart(self):
        self.clear_charts()

        if self.model.rowCount() == 0:
            return

        self.make_chart1()
        self.make_chart2()

    def on_hover(self, point, state):
        if state:
            self.new_mouse_coords.emit(point)

    def clear_charts(self):
        self.chart1.removeAllSeries()
        self.chart2.removeAllSeries()

        if self.chart1.axisX():
            self.chart1.removeAxis(self.chart1.axisX())

        if self.chart1.axisY():
            self.chart1.removeAxis(self.chart1.axisY())

        if self.chart2.axisX():
            self.chart2.removeAxis(self.chart2.axisX())

        if self.chart2.axisY():
            self.chart2.removeAxis(self.chart2.axisY())

File: src/table_portfolio.cpp
#include "table_portfolio.hpp"
#include <QSqlError>
#include <QSqlQuery>
#include "chart_funcs.hpp"

TablePortfolio::TablePortfolio(QWidget* parent) {
  type = TableType::Portfolio;

  name = "portfolio";

  fund_cfg_frame->hide();
  button_add_row->hide();

  connect(spinbox_months, QOverload<int>::of(&QSpinBox::valueChanged), [&](int value) {
    clear_chart(chart2);
    make_chart2();
  });
}

void TablePortfolio::init_model() {
  model->setTable(name);
  model->setEditStrategy(QSqlTableModel::OnManualSubmit);
  model->setSort(1, Qt::DescendingOrder);

  auto currency = QLocale().currencySymbol();

  model->setHeaderData(1, Qt::Horizontal, "Date");
  model->setHeaderData(2, Qt::Horizontal, "Deposit " + currency);
  model->setHeaderData(3, Qt::Horizontal, "Withdrawal " + currency);
  model->setHeaderData(4, Qt::Horizontal, "Starting\n Balance " + currency);
  model->setHeaderData(5, Qt::Horizontal, "Ending\n Balance " + currency);
  model->setHeaderData(6, Qt::Horizontal, "Accumulated\nDeposits " + currency);
  model->setHeaderData(7, Qt::Horizontal, "Accumulated\nWithdrawals " + currency);
  model->setHeaderData(8, Qt::Horizontal, "Net\nDeposit " + currency);
  model->setHeaderData(9, Qt::Horizontal, "Net\nBalance " + currency);
  model->setHeaderData(10, Qt::Horizontal, "Net\nReturn " + currency);
  model->setHeaderData(11, Qt::Horizontal, "Net\nReturn %");
  model->setHeaderData(12, Qt::Horizontal, "Accumulated\nNet Return " + currency);
  model->setHeaderData(13, Qt::Horizontal, "Accumulated\nNet Return %");
  model->setHeaderData(14, Qt::Horizontal, "Real\nReturn %");
  model->setHeaderData(15, Qt::Horizontal, "Accumulated\nReal Return %");

  model->select();

  table_view->setModel(model);
  table_view->setColumnHidden(0, true);
}

void TablePortfolio::process_fund_tables(const QVector<TableFund*>& tables) {
  // get date values in each investment tables

  QSet<int> list_set;

  for (auto& table : tables) {
    // making sure all the latest data was saved to the database

    table->model->submitAll();

    auto query = QSqlQuery(db);

    query.prepare("select distinct date from " + table->name + " order by date");

    if (query.exec()) {
      while (query.next()) {
        list_set.insert(query.value(0).toInt());
      }
    } else {
      qDebug(table->model->lastError().text().toUtf8());
    }
  }

  if (list_set.size() == 0) {
    return;
  }

  QList<int> list_dates = list_set.values();

  std::sort(list_dates.begin(), list_dates.end());

  // calculate columns

  auto qdt = QDateTime();

  for (auto& date : list_dates) {
    double deposit = 0.0, withdrawal = 0.0, starting_balance = 0.0, ending_balance = 0.0, accumulated_deposit = 0.0,
           accumulated_withdrawal = 0.0, net_deposit = 0.0, net_balance = 0.0, net_return = 0.0,
           accumulated_net_return = 0.0;

    qdt.setSecsSinceEpoch(date);

    QString date_month = qdt.toString("MM/yyyy");

    for (auto& table : tables) {
      for (int n = 0; n < table->model->rowCount(); n++) {
        QString tdate = table->model->record(n).value("date").toString();

        if (date_month == QDate::fromString(tdate, "dd/MM/yyyy").toString("MM/yyyy")) {
          deposit += table->model->record(n).value("deposit").toDouble();
          withdrawal += table->model->record(n).value("withdrawal").toDouble();
          starting_balance += table->model->record(n).value("starting_balance").toDouble();
          ending_balance += table->model->record(n).value("ending_balance").toDouble();
          accumulated_deposit += table->model->record(n).value("accumulated_deposit").toDouble();
          accumulated_withdrawal += table->model->record(n).value("accumulated_withdrawal").toDouble();
          net_deposit += table->model->record(n).value("net_deposit").toDouble();
          net_balance += table->model->record(n).value("net_balance").toDouble();
          net_return += table->model->record(n).value("net_return").toDouble();
          accumulated_net_return += table->model->record(n).value("accumulated_net_return").toDouble();

          break;
        }
      }
    }

    double net_return_perc = 100 * net_return / (starting_balance + deposit - withdrawal);

    double real_return_perc = net_return_perc;

    // get inflation values so we can update real_return_perc

    QVector<int> inflation_dates;
    QVector<double> inflation_values, inflation_accumulated;

    auto query = QSqlQuery(db);

    query.prepare("select distinct date,value,accumulated from inflation order by date");

    if (query.exec()) {
      while (query.next()) {
        inflation_dates.append(query.value(0).toInt());
        inflation_values.append(query.value(1).toDouble());
        inflation_accumulated.append(query.value(2).toDouble());
      }
    } else {
      qDebug("Failed to get inflation table values!");
    }

    for (int i = 0; i < inflation_dates.size(); i++) {
      qdt.setSecsSinceEpoch(inflation_dates[i]);

      if (qdt.toString("MM/yyyy") == date_month) {
        real_return_perc = 100.0 * (net_return_perc - inflation_values[i]) / (100.0 + inflation_values[i]);

        break;
      }
    }

    double accumulated_net_return_perc = 0, accumulated_real_return_perc = 0;

    query = QSqlQuery(db);

    query.prepare("insert or replace into " + name + " values ((select id from " + name +
                  " where date == ?),?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)");

    query.addBindValue(date);
    query.addBindValue(date);
    query.addBindValue(deposit);
    query.addBindValue(withdrawal);
    query.addBindValue(starting_balance);
    query.addBindValue(ending_balance);
    query.addBindValue(accumulated_deposit);
    query.addBindValue(accumulated_withdrawal);
    query.addBindValue(net_deposit);
    query.addBindValue(net_balance);
    query.addBindValue(net_return);
    query.addBindValue(net_return_perc);
    query.addBindValue(accumulated_net_return);
    query.addBindValue(accumulated_net_return_perc);
    query.addBindValue(real_return_perc);
    query.addBindValue(accumulated_real_return_perc);

    if (!query.exec()) {
      qDebug(model->lastError().text().toUtf8());
    }
  }

  model->select();

  if (model->rowCount() > 0) {
    calculate_accumulated_sum("net_return");
    calculate_accumulated_product("net_return_perc");
    calculate_accumulated_product("real_return_perc");

    clear_charts();

    make_chart1();
    make_chart2();
  }
}

void TablePortfolio::make_chart1() {
  chart1->setTitle(name.toUpper());

  add_axes_to_chart(chart1, QLocale().currencySymbol());

  auto s1 = add_series_to_chart(chart1, model, "Net Deposit", "net_deposit");
  auto s2 = add_series_to_chart(chart1, model, "Net Balance", "net_balance");
  auto s3 = add_series_to_chart(chart1, model, "Net Return", "accumulated_net_return");

  connect(s1, &QLineSeries::hovered, this,
          [=](const QPointF& point, bool state) { on_chart_mouse_hover(point, state, callout1, s1->name()); });
  connect(s2, &QLineSeries::hovered, this,
          [=](const QPointF& point, bool state) { on_chart_mouse_hover(point, state, callout1, s2->name()); });
  connect(s3, &QLineSeries::hovered, this,
          [=](const QPointF& point, bool state) { on_chart_mouse_hover(point, state, callout1, s3->name()); });
}

void TablePortfolio::make_chart2() {
  chart2->setTitle(name.toUpper());

  add_axes_to_chart(chart2, "%");

  QVector<int> dates;
  QVector<double> net_return, real_return, accumulated_net_return, accumulated_real_return;

  auto query = QSqlQuery(db);

  query.prepare("select distinct date,net_return_perc,real_return_perc from " + name + " order by date desc");

  if (query.exec()) {
    while (query.next() && dates.size() < spinbox_months->value()) {
      dates.append(query.value(0).toInt());
      net_return.append(query.value(1).toDouble());
      real_return.append(query.value(2).toDouble());
    }
  }

  if (dates.size() == 0) {
    return;
  }

  perc_chart_oldest_date = dates[dates.size() - 1];

  std::reverse(dates.begin(), dates.end());
  std::reverse(net_return.begin(), net_return.end());
  std::reverse(real_return.begin(), real_return.end());

  for (int n = 0; n < dates.size(); n++) {
    net_return[n] = net_return[n] * 0.01 + 1.0;
    real_return[n] = real_return[n] * 0.01 + 1.0;
  }

  // cumulative product

  accumulated_net_return.resize(net_return.size());
  accumulated_real_return.resize(real_return.size());

  std::partial_sum(net_return.begin(), net_return.end(), accumulated_net_return.begin(), std::multiplies<double>());
  std::partial_sum(real_return.begin(), real_return.end(), accumulated_real_return.begin(), std::multiplies<double>());

  for (int n = 0; n < dates.size(); n++) {
    accumulated_net_return[n] = (accumulated_net_return[n] - 1.0) * 100;
    accumulated_real_return[n] = (accumulated_real_return[n] - 1.0) * 100;
  }

  auto s1 = add_series_to_chart(chart2, dates, accumulated_net_return, "Net Return");

  connect(s1, &QLineSeries::hovered, this,
          [=](const QPointF& point, bool state) { on_chart_mouse_hover(point, state, callout2, s1->name()); });

  auto s2 = add_series_to_chart(chart2, dates, accumulated_real_return, "Real Return");

  connect(s2, &QLineSeries::hovered, this,
          [=](const QPointF& point, bool state) { on_chart_mouse_hover(point, state, callout2, s2->name()); });

  // ask the main window class for the benchmarks

  emit getBenchmarkTables();
}

void TablePortfolio::show_benchmark(const TableBenchmarks* btable) {
  auto [dates, values, accumulated] = process_benchmark(btable->name, perc_chart_oldest_date);

  if (chart2->axes().size() == 0) {
    return;
  }

  auto series = new QLineSeries();

  series->setName(btable->name.toLower());

  connect(series, &QLineSeries::hovered, this,
          [=](const QPointF& point, bool state) { on_chart_mouse_hover(point, state, callout2, series->name()); });

  double vmin = static_cast<QValueAxis*>(chart2->axes(Qt::Vertical)[0])->min();
  double vmax = static_cast<QValueAxis*>(chart2->axes(Qt::Vertical)[0])->max();

  for (int n = 0; n < dates.size(); n++) {
    auto qdt = QDateTime::fromSecsSinceEpoch(dates[n]);

    auto epoch_in_ms = qdt.toMSecsSinceEpoch();

    double v = accumulated[n];

    if (chart2->series().size() > 0) {
      vmin = std::min(vmin, v);
      vmax = std::max(vmax, v);
    } else {
      if (n == 0) {
        vmin = v;
        vmax = v;
      } else {
        vmin = std::min(vmin, v);
        vmax = std::max(vmax, v);
      }
    }

    series->append(epoch_in_ms, v);
  }

  chart2->addSeries(series);

  series->attachAxis(chart2->axes(Qt::Horizontal)[0]);
  series->attachAxis(chart2->axes(Qt::Vertical)[0]);

  chart2->axes(Qt::Vertical)[0]->setRange(vmin - 0.05 * fabs(vmin), vmax + 0.05 * fabs(vmax));
}

std::tuple<QVector<int>, QVector<double>, QVector<double>> TablePortfolio::process_benchmark(
    const QString& table_name,
    const int& oldest_date) const {
  QVector<int> dates;
  QVector<double> values, accu;

  auto query = QSqlQuery(db);

  query.prepare("select distinct date,value from " + table_name + " where date >= ? order by date");

  query.addBindValue(oldest_date);

  if (query.exec()) {
    while (query.next()) {
      dates.append(query.value(0).toInt());
      values.append(query.value(1).toDouble());
    }
  } else {
    qDebug("Failed to get inflation table values!");
  }

  for (auto& v : values) {
    v = v * 0.01 + 1.0;
  }

  // cumulative product

  accu.resize(values.size());

  std::partial_sum(values.begin(), values.end(), accu.begin(), std::multiplies<double>());

  for (auto& v : accu) {
    v = (v - 1.0) * 100;
  }

  for (auto& v : values) {
    v = (v - 1.0) * 100;
  }

  return {dates, values, accu};
}
